consensus/polybft/block_builder.go

```diff
@@ -0,0 +1,290 @@
+
+	// Vanity extra for the block
+	Extra []byte
+
+	// GasLimit is the gas limit for the block
+	GasLimit uint64
+
+	// duration for one block
+	BlockTime time.Duration
+
+	// Logger
+	//Logger hcf.Logger
+
+	// TxPool txPoolInterface // Reference to the transaction pool
+}
+
+func NewBlockBuilder(params *BlockBuilderParams) *BlockBuilder {
+	// extra can only be 32 size max. it is better to trim that to return
+	// an error that we have to propagate. It should be up to higher level
+	// code to error if the extra supplied by the user is too big.
+	if len(params.Extra) > signer.IstanbulExtraVanity {
+		params.Extra = params.Extra[:signer.IstanbulExtraVanity]
+	}
+
+	if params.Extra == nil {
+		params.Extra = make([]byte, 0)
+	}
+
+	if params.BlockTime == 0 {
+		params.BlockTime = time.Second * 2 // TODO: is this ok?
+	}
+
+	builder := &BlockBuilder{
+		params: params,
+	}
+	builder.Reset()
+
+	return builder
+}
+
+var _ blockBuilder = &BlockBuilder{}
+
+type BlockBuilder struct {
+	// input params for the block
+	params *BlockBuilderParams
+
+	// header is the header for the block being created
+	header *types.Header
+
+	// transactions are the data included in the block
+	txns []*types.Transaction
+
+	// block is a reference to the already built block
+	block *types.Block
+
+	state *state.Transition
+}
+
+// Reset is used to indicate that the current block building has been interrupted
+// and it has to clean any data
+func (b *BlockBuilder) Reset() {
+	b.header = &types.Header{
+		ParentHash:   b.params.Parent.Hash,
+		Number:       b.params.Parent.Number + 1,
+		Miner:        b.params.Coinbase[:],
+		Difficulty:   1,
+		ExtraData:    b.params.Extra,
+		StateRoot:    types.EmptyRootHash, // this avoids needing state for now
+		TxRoot:       types.EmptyRootHash,
+		ReceiptsRoot: types.EmptyRootHash, // this avoids needing state for now
+		Sha3Uncles:   types.EmptyUncleHash,
+		GasLimit:     b.params.GasLimit, //will need to adjust dynamically later.
+	}
+
+	b.block = nil
+	b.txns = []*types.Transaction{}
+
+	transition, err := b.params.Executor.BeginTxn(b.params.Parent.StateRoot, b.header, b.params.Coinbase)
+	if err != nil {
+		panic(err)
+	}
+
+	b.state = transition
+}
+
+// Block returns the built block if nil, it is not built yet
+func (b *BlockBuilder) Block() *types.Block {
+	return b.block
+}
+
+// Build creates the state and the final block
+func (b *BlockBuilder) Build(handler func(h *types.Header)) (*StateBlock, error) {
+	if handler != nil {
+		handler(b.header)
+	}
+
+	// TODO: Nemanja - see what to do with gas
+	// calculate gas limit based on parent header
+	//gasLimit, err := b.blockchain.CalculateGasLimit(header.Number)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// header.GasLimit = gasLimit
+
+	// Nemanja - fill transactions
+	// txs := i.writeTransactions(gasLimit, header.Number, transition)
+
+	// if err := i.PreCommitState(header, transition); err != nil {
+	// 	return nil, err
+	// }
```

Comment:
Leftover?

Reply:
Changed in another https://github.com/0xPolygon/polygon-edge/pull/789